From d70b592551ff23747e26e74081205babf8dba9b6 Mon Sep 17 00:00:00 2001
From: David Bauer <mail@david-bauer.net>
Date: Tue, 18 Nov 2025 01:16:18 +0100
Subject: [PATCH] l2tp: reset skb control buffer on xmit

[ Upstream commit d70b592551ff23747e26e74081205babf8dba9b6 ]

The L2TP stack did not reset the skb control buffer before sending the
encapsulated package.

In a setup with an ath10k radio and batman-adv over an L2TP tunnel
massive fragmentations happen sporadically if the L2TP tunnel is
established over IPv4.

L2TP might reset some of the fields in the IP control buffer, but L2TP
assumes the type of the control buffer to be of an IPv4 packet.

In case the L2TP interface is used as a batadv hardif or the packet is
an IPv6 packet, this assumption breaks.

Clear the entire control buffer to avoid such mishaps altogether.

Fixes: f77ae9390438 ("[PPPOL2TP]: Reset meta-data in xmit function")
Signed-off-by: David Bauer <mail@david-bauer.net>
Link: https://patch.msgid.link/20251118001619.242107-1-mail@david-bauer.net
Signed-off-by: Paolo Abeni <pabeni@redhat.com>

[ Neanderfunk backport to Linux 4.14.275 (OpenWrt 19.07, Gluon v2021.1),
  following Gluon b7f250df (v2023.2.x) / v2025.1.x
  patches/openwrt/0011-net-l2tp-reset-skb-control-buffer.patch:
  - In 4.14 the reset is in l2tp_xmit_skb() right after
    session->build_header(); l2tp_xmit_core() does not build the header yet.
    The memset is placed at the same point (after the L2TP header, before
    the UDP header and ip_queue_xmit()), as upstream.
  - 4.14 still uses nf_reset() (renamed to nf_reset_ct() in 5.4); the call
    is kept unchanged.
  - The IPCB flags line is wrapped differently in 4.14, so the removed lines
    differ from upstream. Nothing on the 4.14 transmit path reads skb->cb
    between this point and ip_queue_xmit(); L2TP_SKB_CB() is only used on
    receive. ]
---
 net/l2tp/l2tp_core.c |    7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/net/l2tp/l2tp_core.c b/net/l2tp/l2tp_core.c
--- a/net/l2tp/l2tp_core.c
+++ b/net/l2tp/l2tp_core.c
@@ -1195,10 +1195,9 @@ int l2tp_xmit_skb(struct l2tp_session *s
 	/* Setup L2TP header */
 	session->build_header(session, __skb_push(skb, hdr_len));
 
-	/* Reset skb netfilter state */
-	memset(&(IPCB(skb)->opt), 0, sizeof(IPCB(skb)->opt));
-	IPCB(skb)->flags &= ~(IPSKB_XFRM_TUNNEL_SIZE | IPSKB_XFRM_TRANSFORMED |
-			      IPSKB_REROUTED);
+	/* Reset control buffer */
+	memset(skb->cb, 0, sizeof(skb->cb));
+
 	nf_reset(skb);
 
 	bh_lock_sock(sk);
